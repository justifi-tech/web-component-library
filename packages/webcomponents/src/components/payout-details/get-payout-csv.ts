import { ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetPayoutCSV =
  ({ authToken, service }) =>
  async ({ payoutId, onError }) => {
    try {
      const { data } = await service.fetchCSV(payoutId, authToken);
      const a = document.createElement('a');
      a.href = data.csv_url;
      a.click();
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      const code = getErrorCode(error?.code);
      onError({
        error: errorMessage,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };
