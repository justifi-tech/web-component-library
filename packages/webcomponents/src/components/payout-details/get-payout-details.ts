import { Payout } from '../../api';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetPayoutDetails =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchPayout(id, authToken);

      if (!response.error) {
        const payout = new Payout(response.data);

        onSuccess(payout);
      } else {
        const responseError = getErrorMessage(response.error);
        const errorMessage = responseError;
        onError(getErrorMessage(errorMessage));
      }
    } catch (error) {
      const errorMessage = error;
      onError(getErrorMessage(errorMessage));
    }
  };
