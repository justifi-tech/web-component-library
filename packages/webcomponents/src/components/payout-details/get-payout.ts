import { Payout } from '../../api';

export const makeGetPayout =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchPayout(id, authToken);

      if (!response.error) {
        const payout = new Payout(response.data);

        onSuccess(payout);
      } else {
        const responseError =
          typeof response.error === 'string'
            ? response.error
            : response.error.message;
        const errorMessage = `Error fetching payout details: ${responseError}`;
        onError(errorMessage);
      }
    } catch (error) {
      const errorMessage = `Error fetching payout details: ${error}`;
      onError(errorMessage);
    }
  };
