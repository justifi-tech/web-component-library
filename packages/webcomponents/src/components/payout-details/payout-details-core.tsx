import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { IPayout, Payout } from '../../api';
import { formatCurrency, formatDate, formatTime } from '../../utils/utils';
import { CodeBlock, DetailItem, DetailSectionTitle, EntityHeadInfo, EntityHeadInfoItem, ErrorState, LoadingState } from '../../ui-components/details/utils';
import { ComponentError } from '../../api/ComponentError';
import { StyledHost } from '../../ui-components';
import { MapPayoutStatusToBadge } from '../payouts-list/payouts-status';

@Component({
  tag: 'payout-details-core',
})

export class PayoutDetailsCore {
  @Prop() getPayout: Function;
  @Prop() getPayoutCSV: Function;

  @State() payout: Payout;
  @State() loading: boolean = true;
  @State() errorMessage: string = null;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    if (this.getPayout) {
      this.fetchData();
    }
  }

  @Watch('getPayout')
  getPayoutWatcher() {
    this.fetchData();
  }

  fetchData(): void {
    this.loading = true;

    this.getPayout({
      onSuccess: (payout: IPayout) => {
        this.errorMessage = null;
        this.payout = payout;
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          message: error,
          errorCode: code,
          severity,
        });
        this.loading = false;
      },
    });
  }

  downloadCSV = () => {
    this.getPayoutCSV({
      payoutId: this.payout.id,
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          message: error,
          errorCode: code,
          severity,
        });
      },
    });
  }

  render() {
    return (
      <StyledHost>
        {this.loading && LoadingState()}
        {!this.loading && this.errorMessage && ErrorState(this.errorMessage)}
        {!this.loading && this.payout && (
          <justifi-details error-message={this.errorMessage}>
            <EntityHeadInfo
              slot="head-info"
              badge={<span slot='badge' part="detail-head-badge" innerHTML={MapPayoutStatusToBadge(this.payout?.status)} />}
              title={`${formatCurrency(this.payout.amount)} ${this.payout.currency.toUpperCase()}`}
            >
              <EntityHeadInfoItem
                classes="border-1 border-end"
                title="Updated At"
                value={`${formatDate(this.payout.updated_at)} ${formatTime(this.payout.updated_at)}`}
              />
              <EntityHeadInfoItem
                classes="border-1 border-end"
                title="Created At"
                value={`${formatDate(this.payout.created_at)} ${formatTime(this.payout.created_at)}`}
              />
              <EntityHeadInfoItem title="ID" value={this.payout.id} />
              <div class="m-4">
                <button part="button-secondary" class="btn btn-outline-secondary d-flex align-items-center" onClick={this.downloadCSV}>
                  Export CSV
                </button>
              </div>
            </EntityHeadInfo>
            <div slot='detail-sections' part="detail-section">
              <DetailSectionTitle sectionTitle="Details" />
              <div class="d-table gap-2 w-100">
                <DetailItem title="Date paid" value={formatDate(this.payout.deposits_at)} />
                <DetailItem title="Statement Description" value={this.payout.description} />
                <DetailItem title="Payout Method" value={this.payout.delivery_method} />
                <DetailItem title="Amount" value={formatCurrency(this.payout.amount)} />
                <DetailItem title="Fee" value={formatCurrency(this.payout.fees_total)} />
              </div>
              <DetailSectionTitle sectionTitle="Account" />
              <div class="d-table gap-2 w-100">
                <DetailItem title="ID" value={this.payout.account_id} />
                <DetailItem title="Account Type" value={this.payout.bank_account.account_type} />
                <DetailItem title="Institution" value={this.payout.bank_account.account_type} />
                <DetailItem title="Routing Number" value={this.payout.bank_account.routing_number} />
                <DetailItem title="Account Number" value={this.payout.bank_account.account_number_last4} />
              </div>
              <DetailSectionTitle sectionTitle='Metadata' />
              <div class="d-table gap-2 w-100">
                <CodeBlock metadata={this.payout.metadata} />
              </div>
            </div>
          </justifi-details>
        )}
      </StyledHost>
    )
  }
}
