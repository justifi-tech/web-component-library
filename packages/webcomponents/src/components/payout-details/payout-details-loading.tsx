import { h } from "@stencil/core";
import { Skeleton } from "../../ui-components";

const PayoutDetailsLoading = () => (
  <div class="row gap-3">
    <div class="col-12"><Skeleton width="35%" height="30px" /></div>
    <div class="col-12"><Skeleton width="70%" /></div>
    <div class="col-12"><Skeleton width="40%" /></div>
    <div class="col-12"><Skeleton width="40%" /></div>
    <div class="col-12"><Skeleton width="100%" /></div>
    <div class="col-12"><Skeleton width="80%" /></div>
  </div>
)

export default PayoutDetailsLoading;
