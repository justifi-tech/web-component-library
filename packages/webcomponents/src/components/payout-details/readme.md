# justifi-payout-details



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute | Description | Type       | Default     |
| -------------- | --------- | ----------- | ---------- | ----------- |
| `getPayout`    | --        |             | `Function` | `undefined` |
| `getPayoutCSV` | --        |             | `Function` | `undefined` |


## Events

| Event         | Description | Type                               |
| ------------- | ----------- | ---------------------------------- |
| `error-event` |             | `CustomEvent<ComponentErrorEvent>` |


## Dependencies

### Used by

 - [justifi-payout-details](.)

### Depends on

- [justifi-details](../../ui-components/details)

### Graph
```mermaid
graph TD;
  payout-details-core --> justifi-details
  justifi-payout-details --> payout-details-core
  style payout-details-core fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
