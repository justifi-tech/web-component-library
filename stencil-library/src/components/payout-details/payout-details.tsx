import { Component, Host, h, Prop, State, Watch } from '@stencil/core';
import { Api, IApiResponseCollection, Payout } from '../../api';
import { MapPayoutStatusToBadge, formatCurrency, formatDate, formatTime } from '../../utils/utils';
import { CodeBlock, DetailItem, DetailSection, EntityHeadInfo, EntityHeadInfoItem, ErrorState, LoadingState } from '../details/utils';
import { config } from '../../../config';

/**
  * @exportedPart detail-loading-spinner
  * @exportedPart detail-loading-state
  * @exportedPart detail-empty-state
  * @exportedPart detail-head
  * @exportedPart detail-title
  * @exportedPart detail-method
  * @exportedPart detail-info
  * @exportedPart detail-info-item
  * @exportedPart detail-info-item-title
  * @exportedPart detail-info-item-data
  * @exportedPart detail-metadata
  * @exportedPart detail-metadata-title
  * @exportedPart detail-method-title
  * @exportedPart detail-method-data
  * @exportedPart detail-section
  * @exportedPart detail-section-title
  * @exportedPart detail-section-item-title
  * @exportedPart detail-section-item-data
  * @exportedPart detail-head-info
  * @exportedPart detail-head-info-item
  * @exportedPart detail-head-info-item-title
  * @exportedPart detail-head-info-item-data
*/
@Component({
  tag: 'justifi-payout-details',
  styleUrl: 'payout-details.scss',
  shadow: true,
})

export class PaymentDetails {
  @Prop() payoutId: string;
  @Prop() authToken: string;
  @State() payout: Payout;
  @State() loading: boolean = true;
  @State() errorMessage: string;

  @Watch('payoutId')
  @Watch('authToken')
  updateOnPropChange() {
    this.fetchData();
  }

  connectedCallback() {
    this.fetchData();
  }

  async fetchData(): Promise<void> {
    this.errorMessage = '';
    if (!this.payoutId || !this.authToken) {
      this.errorMessage = "Can not fetch any data without a PayoutID and an AuthToken";
      this.loading = false;
      return;
    }
    this.loading = true;
    const endpoint = `payouts/${this.payoutId}`;

    const response: IApiResponseCollection<Payout> = await Api(this.authToken, config.proxyApiOrigin).get(endpoint);
    if (!response.error) {
      this.payout = response.data;
    } else {
      this.errorMessage = typeof response.error === 'string' ? response.error : response.error.message;
    }

    this.loading = false;
  }

  render() {
    return (
      <Host>
        {this.loading && LoadingState}
        {!this.loading && this.errorMessage && ErrorState(this.errorMessage)}
        {!this.loading && this.payout && (
          <justifi-details error-message={this.errorMessage}>
            <EntityHeadInfo slot="head-info" badge={<span slot='badge' innerHTML={MapPayoutStatusToBadge(this.payout?.status)} />} title={`${formatCurrency(this.payout.amount)} ${this.payout.currency.toUpperCase()}`}>
              <EntityHeadInfoItem
                classes="border-1 border-end"
                title="Updated At"
                value={`${formatDate(this.payout.updated_at)} ${formatTime(this.payout.updated_at)}`}
              />
              <EntityHeadInfoItem
                classes="border-1 border-end"
                title="Created At"
                value={`${formatDate(this.payout.created_at)} ${formatTime(this.payout.created_at)}`}
              />
              <EntityHeadInfoItem title="ID" value={this.payout.id} />
            </EntityHeadInfo>
            <div slot='detail-sections'>
              <DetailSection sectionTitle="Details">
                <DetailItem title="Date paid" value={formatDate(this.payout.deposits_at)} />
                <DetailItem title="Statement Description" value={this.payout.description} />
                <DetailItem title="Payout Method" value={this.payout.delivery_method} />
                <DetailItem title="Amount" value={formatCurrency(this.payout.amount)} />
                <DetailItem title="Fee" value={formatCurrency(this.payout.fees_total)} />
              </DetailSection>
              <DetailSection sectionTitle="Account">
                <DetailItem title="ID" value={this.payout.account_id} />
                <DetailItem title="Account Type" value={this.payout.bank_account.account_type} />
                <DetailItem title="Institution" value={this.payout.bank_account.account_type} />
                <DetailItem title="Routing Number" value={this.payout.bank_account.routing_number} />
                <DetailItem title="Account Number" value={this.payout.bank_account.account_number_last4} />
              </DetailSection>
              <DetailSection sectionTitle='Metadata'>
                <CodeBlock metadata={this.payout.metadata} />
              </DetailSection>
            </div>
          </justifi-details>
        )}
      </Host>
    )
  }
}