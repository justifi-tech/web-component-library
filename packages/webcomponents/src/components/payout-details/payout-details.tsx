import { Component, h, Prop, Watch, State, Event, EventEmitter } from '@stencil/core';
import { PayoutService } from '../../api/services/payout.service';
import { makeGetPayoutDetails } from './get-payout-details';
import { ErrorState } from '../details/utils';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';
import JustifiAnalytics from '../../api/Analytics';
import { makeGetPayoutCSV } from './get-payout-csv';

@Component({
  tag: 'justifi-payout-details',
  shadow: true,
})

export class PayoutDetails {
  @Prop() payoutId: string;
  @Prop() authToken: string;

  @State() getPayout: Function;
  @State() getPayoutCSV: Function;
  @State() errorMessage: string = null;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  analytics: JustifiAnalytics;

  componentWillLoad() {
    this.analytics = new JustifiAnalytics(this);
    this.initializeServices();
  }

  disconnectedCallback() {
    this.analytics.cleanup();
  }

  @Watch('payoutId')
  @Watch('authToken')
  propChanged() {
    this.initializeServices();
  }

  initializeServices() {
    if (this.payoutId && this.authToken) {
      this.getPayout = makeGetPayoutDetails({
        id: this.payoutId,
        authToken: this.authToken,
        service: new PayoutService()
      });
      this.getPayoutCSV = makeGetPayoutCSV({
        authToken: this.authToken,
        service: new PayoutService()
      });
    } else {
      this.errorMessage = 'Failed to load payout details. payoutId or authToken is not provided.';
      this.errorEvent.emit({
        message: this.errorMessage,
        errorCode: ComponentErrorCodes.MISSING_PROPS,
        severity: ComponentErrorSeverity.ERROR
      });
    }
  }

  handleErrorEvent = (event: CustomEvent<ComponentError>) => {
    this.errorMessage = event.detail.message;
    this.errorEvent.emit(event.detail);
  };

  render() {
    if (this.errorMessage) {
      return ErrorState(this.errorMessage);
    }

    return (
      <payout-details-core
        getPayout={this.getPayout}
        getPayoutCSV={this.getPayoutCSV}
        onError-event={this.handleErrorEvent}
      />
    );
  }
}
