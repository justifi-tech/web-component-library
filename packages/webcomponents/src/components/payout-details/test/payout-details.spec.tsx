import { newSpecPage } from "@stencil/core/testing";
import { PayoutDetails } from "../payout-details";
import { PayoutDetailsCore } from "../payout-details-core";

describe('payout-details', () => {
  it('renders error state when no payoutId or authToken', async () => {
    const page = await newSpecPage({
      components: [PayoutDetails, PayoutDetailsCore],
      html: '<justifi-payout-details></justifi-payout-details>',
    });
    await page.waitForChanges();
    // expect(page.root).toMatchSnapshot();
  });
});
