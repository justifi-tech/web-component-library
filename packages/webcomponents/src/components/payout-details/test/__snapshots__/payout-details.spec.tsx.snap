// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payout-details renders error state when no payoutId or authToken 1`] = `
<justifi-payout-details>
  <template shadowrootmode="open">
    <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
      <div>
        Failed to load payout details. payoutId or authToken is not provided.
      </div>
    </main>
  </template>
</justifi-payout-details>
`;
